#104 Run tests for the user creating the PR

filiphr opened the pull request:
Adding a step in the GitHub action to test the submission of the user submitting the PR

Fixes #104

@gunnarmorling please wait for merging, I want to also push some failures to see how it looks like.

cc @AlexanderYastrebov

filiphr commented:
A failure would look like

![image](https://github.com/gunnarmorling/1brc/assets/6012094/4f6e71eb-4972-4d65-8d35-109e5b750e80)

and then if you go into details you'll see

![image](https://github.com/gunnarmorling/1brc/assets/6012094/c12b383b-77a0-4a33-a700-c6e3302ab83a)

filiphr commented:
I've also added a check to validate that the file exists. If a file does not exists it will have an error like

![image](https://github.com/gunnarmorling/1brc/assets/6012094/a9b33f6c-e6d5-4ab8-8e2f-381684156ff7)

filiphr commented:
@gunnarmorling this is now ready for merging. I've removed the commits that were causing failures.

AlexanderYastrebov commented on test.sh:
```diff
@@ -22,12 +22,18 @@ if [ -z "$1" ]; then
+fi
+
 for sample in $(ls src/test/resources/samples/*.txt); do
   echo "Validating calculate_average_$1.sh -- $sample"
```
I guess we can change this to `echo "Validating $1 -- $sample"` to make it a bit cleaner

filiphr replied:
Sounds OK to me. I'll change it.

filiphr replied:
I forgot to change this. I think we can merge without it, and we can change it later.

AlexanderYastrebov commented on .github/workflows/maven.yml:
```diff
@@ -47,3 +47,7 @@ jobs:
         run: mvn -B clean verify -Pci
+
+      - name: 'Test submission'
+        run: ./test.sh ${{ github.event.pull_request.user.login }}
```
I am not sure this will work for all submissions, e.g. some use sdk or graal

filiphr replied:
Indeed it might not work. I'll see if we can use https://github.com/sdkman/sdkman-action/tree/main

gunnarmorling replied:
Yeah, I think this would be good. Wanna add it to this PR?

filiphr replied:
Was already working on it, but I am travelling today so a bit slow :)

AlexanderYastrebov replied:
There is a related ticket https://github.com/gunnarmorling/1brc/issues/150

AlexanderYastrebov commented:
I thin PR description fixes wrong ticket, should be #104

filiphr commented:
>I thin PR description fixes wrong ticket, should be https://github.com/gunnarmorling/1brc/issues/104

good point. I've updated that

filiphr commented:
I've squashed all the commits and I've added SDKMAN as an action.

For this to work on the CI people will need to do something like:

```
source "$HOME/.sdkman/bin/sdkman-init.sh"
sdk install java 21.0.1-graal 1>&2
sdk use java 21.0.1-graal 1>&2
```

The `sdk install` is needed because the CI will not have the needed JDK. If the specified JDK is installed then `sdk install` is a noop (won't faill).


-----

Aside from this approach, I have also been looking into an alternative. However, it would require a change in all scripts.

The calculate script can look like

```sh
function java_to_use() {
    echo "21.0.1-graal"
}

function run_submission() {
  JAVA_OPTS=""
  time java $JAVA_OPTS --class-path target/average-1.0.0-SNAPSHOT.jar dev.morling.onebrc.CalculateAverage_<username>
}
```
The `java_to_use` is optional and if omitted then the default one is used. We would then have a `helper.sh` which looks like

```sh
source $HOME/.sdkman/bin/sdkman-init.sh

function sdk_use_java() {
  if [ -z "$1" ]; then
    echo "Usage: sdk_use_java <candidate>"
    exit 1
  fi
  java_to_use="$1"
  sdk install java ${java_to_use} 1>&2
  sdk use java ${java_to_use} 1>&2
}

if [ -n "$(LC_ALL=C type -t java_to_use)" ] && [ "$(LC_ALL=C type -t java_to_use)" = function ]; then
  sdk_java=$(java_to_use)
  sdk_use_java ${sdk_java}
fi

java -version
```

the `evaluate.sh` will need to then have before the loop

```sh
source ${CALCULATE_FILE}
source ./helper.sh
```

and in the loop the diff would be

```sh
diff <(run_submission) ${sample%.txt}.out
```

Similar chnage would be needed for the `evaluate.sh` and we would most likelt need to add a `run.sh` that would then run the `calculate` as well.

Let me know what you think about this idea.

gunnarmorling commented:
Could we somehow pre-install all the (currently) used distributions?

gunnarmorling commented:
> Could we somehow pre-install all the (currently) used distributions?

Any thoughts on this, @filiphr? Or should we close it for now?

filiphr commented:
Yes we can do that. I've been away with a really bad internet connection. I'm back on Sunday, will be able to look into it then

gunnarmorling commented:
Ok, closing for now. Thx for looking into this, @filiphr!

gunnarmorling commented:
Gave this another try, see https://github.com/gunnarmorling/1brc/pull/535.
